frontend/src/app/workspace/service/copilot/texera-copilot.ts

```diff
@@ -0,0 +1,388 @@
+      if (!this.model) {
+        return throwError(() => new Error("Copilot not initialized"));
+      }
+
+      if (this.state !== CopilotState.AVAILABLE) {
+        return throwError(() => new Error(`Cannot send message: agent is ${this.state}`));
+      }
+
+      this.setState(CopilotState.GENERATING);
+
+      // Generate unique message ID for this conversation turn
+      const messageId = crypto.randomUUID();
+      let stepId = 0;
+
+      // Emit user message as first step
+      this.emitReActStep(messageId, stepId++, "user", message, true, true);
+      this.messages.push({ role: "user", content: message });
+
+      let isFirstStep = true;
+
+      /**
+       * Generate text using the AI model with ReAct (Reasoning + Acting) pattern.
+       * This is the core of the agent lifecycle with several callbacks:
+       *
+       * Lifecycle flow:
+       * 1. generateText() starts the LLM generation
+       * 2. stopWhen() - checked before each step to determine if generation should stop
+       * 3. onStepFinish() - called DURING generation after each reasoning/action step (real-time updates)
+       * 4. pipe operators - executed AFTER generation completes (final processing)
+       */
+      return from(
+        generateText({
+          model: this.model,
+          messages: this.messages,
+          tools: this.tools,
+          system: COPILOT_SYSTEM_PROMPT,
+          /**
+           * stopWhen - Determines if generation should stop.
+           * Called before each step during generation.
+           * Returns true to stop, false to continue.
+           */
+          stopWhen: ({ steps }) => {
+            if (this.state === CopilotState.STOPPING) {
+              this.notificationService.info(`Agent ${this.agentName} has stopped generation`);
+              return true;
+            }
+            // Stop if step count reaches max limit to prevent infinite loops
+            return stepCountIs(TexeraCopilot.MAX_REACT_STEPS)({ steps });
+          },
+          /**
+           * onStepFinish is called DURING generation after each ReAct step completes.
+           * This provides real-time updates to the UI as the agent reasons and acts.
+           *
+           * Each step may include:
+           * - text: The agent's reasoning or response text
+           * - toolCalls: Tools the agent decided to call
+           * - toolResults: Results from executed tools
+           * - usage: Token usage for this step
+           *
+           * Note: This is called multiple times during a single generation,
+           * once per reasoning/action cycle.
+           */
+          onStepFinish: ({ text, toolCalls, toolResults, usage }) => {
+            if (this.state === CopilotState.STOPPING) {
+              return;
+            }
+
+            // Parse operator access from tool results to track viewed/modified operators
+            const operatorAccess = parseOperatorAccessFromStep(toolCalls || [], toolResults || []);
```

Comment:
This PR would not have actions that modify operators, right?

Reply:
Yes, this function also parses the operator that are being red